cat/opencl.cat

```diff
@@ -0,0 +1,133 @@
+OpenCL
+(* OpenCL Memory Model *)
+include "basic.cat"
+
+(*
+* This model is based on:
+* https://multicore.doc.ic.ac.uk/overhauling/opencl_base.cat
+* https://multicore.doc.ic.ac.uk/overhauling/opencl_scopedsc.cat
+*)
+
+// Base relations:
+// int: same work-item
+// swg: same work-group
+// sdv: same device
+// syncbar: same barrier id
+
+// Tags:
+// WI: work-item scope
+// WG: work-group scope
+// DV: device scope
+// ALL: all-svm-devices scope
+// GLOBAL: global memory
+// LOCAL: local memory
+// EF: entry fence of barrier
+// XF: exit fence of barrier
+
+// dynamic_tag relates events to itself that access an address whose init event is marked X or Fence tagged with X
+let dynamic_tag(X) = [range([IW & X]; loc)] | [X & F]
+
+let mo = dynamic_tag(~NAL); co; dynamic_tag(~NAL)
+let sb = po
+let rb = rf^-1;mo | ([R] \ [range(rf)]);loc;[W]
+let unv = _ * _
+let wi = int
+
+(* Inclusive scopes *)
+let incl = (swg & (WG * WG)) | (sdv & (DV * DV)) | (ALL * ALL)
+
+(*******************)
+(* Synchronisation *)
+(*******************)
+
+let Acq = (ACQ | SC | ACQ_REL) & (R | F)
+let Rel = (REL | SC | ACQ_REL) & (W | F)
+
+(* Fences sequenced before or after *)
+let Fsb = [F]; sb
+let sbF = sb; [F]
+
+(* Release sequence *)
+let rs_prime = wi | (unv; [R & W])
```

Comment:
You can use `wi | _ * (R & W)`. If you prefer the `unv` version you can keep it. My suggestion is just optimized for internal computation (RA is better on non-compositions).
Btw. we do not have events that are `R & W` so this is always empty, no? What's the purpose then?

Reply:
Updated, it was for RMW, I forgot to change it when porting from the herd version.